File: Games/boardgame.py
# File = robjam1990/Psychosis/Gameplay/Games/boardgame.py
import logging
import socketio

logging.basicConfig(level=logging.INFO)
log = logging.getLogger('psychosis_bot')

class Namespace(socketio.ClientNamespace):

    def __init__(self, bot, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.bot = bot
        self.previous_state_id = None

    def on_connect(self):
        log.info('Connected to server')

    def on_disconnect(self):
        log.info('Disconnected from server')

    def on_reconnect(self):
        log.info('Reconnected to server')

    def on_update(self, game_id, state):
        state_id = state.get('_stateID')
        ctx = state.get('ctx')

        if not (state_id and ctx):
            log.warning("Received incomplete game state: %s", state)
            return

        if game_id == self.bot.game_id:
            if not self.previous_state_id or state_id >= self.previous_state_id:
                self.previous_state_id = state_id
                log.debug('Received updated game state: %s', state)
                if 'gameover' in ctx:
                    self.bot.gameover(state['G'], ctx)
                elif self.bot.player_id in ctx.get('actionPlayers', []):
                    log.info('Current game phase: %s', ctx.get('phase'))
                    actions = self.bot.think(state['G'], ctx)
                    if not isinstance(actions, list):
                        actions = [actions]
                    for action in actions:
                        log.info('Sending action: %s', action.get('payload'))
                        self.emit('update', action, state_id, game_id, self.bot.player_id)

class Bot:

    def __init__(self, server='localhost', port='8000', options=None):
        opts = {
            'game_name': 'default',
            'game_id': 'default',
            'player_id': '1',
            'credentials': 'default',
            'num_players': 2
        }
        opts.update(options or {})
        self.game_id = f"{opts['game_name']}:{opts['game_id']}"
        self.player_id = opts['player_id']
        self.credentials = opts['credentials']
        self.num_players = opts['num_players']

        self.sio = socketio.Client()
        self.sio.register_namespace(Namespace(self))
        self.sio.connect(f"http://{server}:{port}")

    def _create_action(self, action, typ, args=None):
        args = args or []
        return {
            'type': action,
            'payload': {
                'type': typ,
                'args': args,
                'playerID': self.player_id,
                'credentials': self.credentials
            }
        }

    def make_move(self, typ, *args):
        return self._create_action('MAKE_MOVE', typ, list(args))

    def game_event(self, typ):
        return self._create_action('GAME_EVENT', typ)

    def think(self, _G, _ctx):
        raise NotImplementedError("Method 'think' must be implemented in a subclass")

    def gameover(self, _G, _ctx):
        raise NotImplementedError("Method 'gameover' must be implemented in a subclass")

    def listen(self, timeout=1):
        try:
            self.sio.wait(seconds=timeout)
        except KeyboardInterrupt:
            log.info("KeyboardInterrupt received. Exiting...")
        except Exception as e:
            log.exception("An error occurred while listening for updates: %s", e)


if __name__ == "__main__":
    # Example subclass implementation of Bot
    class MyBot(Bot):

        def think(self, _G, _ctx):
            # Example: Bot always makes the same move
            return self.make_move('MOVE_TYPE', 'arg1', 'arg2')

        def gameover(self, _G, _ctx):
            log.info("Game over. Final state: %s", _G)

    # Initialize and run the bot
    my_bot = MyBot()
    my_bot.listen()

File: Games/boardgame.js
const io = require("socket.io-client");
const { EventEmitter } = require("events");

const log = console;

class Namespace extends EventEmitter {
    constructor(bot, ...args) {
        super(...args);
        this.bot = bot;
        this.previousStateId = null;
    }

    onConnect() {
        log.info('Connected to server');
    }

    onDisconnect() {
        log.info('Disconnected from server');
    }

    onReconnect() {
        log.info('Reconnected to server');
    }

    onUpdate(gameId, state) {
        const stateId = state._stateID;
        const ctx = state.ctx;

        if (!(stateId && ctx)) {
            log.warning("Received incomplete game state: ", state);
            return;
        }

        if (gameId === this.bot.gameId) {
            if (!this.previousStateId || stateId >= this.previousStateId) {
                this.previousStateId = stateId;
                log.debug('Received updated game state: ', state);
                if ('gameover' in ctx) {
                    this.bot.gameover(state.G, ctx);
                } else if (ctx.actionPlayers.includes(this.bot.playerId)) {
                    log.info('Current game phase: ', ctx.phase);
                    const actions = this.bot.think(state.G, ctx);
                    if (!Array.isArray(actions)) {
                        actions = [actions];
                    }
                    for (const action of actions) {
                        log.info('Sending action: ', action.payload);
                        this.emit('update', action, stateId, gameId, this.bot.playerId);
                    }
                }
            }
        }
    }
}

class Bot {
    constructor(server = 'localhost', port = '8000', options = null) {
        const opts = {
            game_name: 'default',
            game_id: 'default',
            player_id: '1',
            credentials: 'default',
            num_players: 2
        };
        Object.assign(opts, options || {});
        this.gameId = `${opts.game_name}:${opts.game_id}`;
        this.playerId = opts.player_id;
        this.credentials = opts.credentials;
        this.numPlayers = opts.num_players;

        this.sio = io.connect(`http://${server}:${port}`);
        this.namespace = new Namespace(this);
        this.sio.on('connect', () => this.namespace.onConnect());
        this.sio.on('disconnect', () => this.namespace.onDisconnect());
        this.sio.on('reconnect', () => this.namespace.onReconnect());
        this.sio.on('update', (gameId, state) => this.namespace.onUpdate(gameId, state));
    }

    _createAction(action, typ, args = []) {
        return {
            type: action,
            payload: {
                type: typ,
                args: args,
                playerID: this.playerId,
                credentials: this.credentials
            }
        };
    }

    makeMove(typ, ...args) {
        return this._createAction('MAKE_MOVE', typ, args);
    }

    gameEvent(typ) {
        return this._createAction('GAME_EVENT', typ);
    }

    think(_G, _ctx) {
        throw new Error("Method 'think' must be implemented in a subclass");
    }

    gameover(_G, _ctx) {
        throw new Error("Method 'gameover' must be implemented in a subclass");
    }

    listen(timeout = 1) {
        try {
            this.sio.connect();
        } catch (error) {
            if (error instanceof KeyboardInterrupt) {
                log.info("KeyboardInterrupt received. Exiting...");
            } else {
                log.exception("An error occurred while listening for updates: ", error);
            }
        }
    }
}

if (require.main === module) {
    // Example subclass implementation of Bot
    class MyBot extends Bot {
        think(_G, _ctx) {
            // Example: Bot always makes the same move
            return this.makeMove('MOVE_TYPE', 'arg1', 'arg2');
        }

        gameover(_G, _ctx) {
            log.info("Game over. Final state: ", _G);
        }
    }

    // Initialize and run the bot
    const myBot = new MyBot();
    myBot.listen();
}
